### ISSUE TITLE
C.21 Exception for copy-and-swap idiom

### ISSUE THREAD

**Issue by Eisenwave** [isocpp/CppCoreGuidelines#2164]
C.21 is missing an exception for the copy-and-swap idiom. It's perfectly safe (and considered idiomatic by many) for resource management to write:
```cpp
T& operator=(T) noexcept { /* ... */ }
```

This would violate C.21 since technically, the user has only defined the copy assignment operator, and the move assignment operator is implicitly not declared.

The lack of an exception to this rule also manifests itself in clang-tidy false positives. See:
- https://github.com/llvm/llvm-project/issues/76064
- https://stackoverflow.com/questions/77692186/clang-tidy-does-not-recognize-unifying-assignment-operator-as-move-assigment

**Comment by sergio-nsk** [comment]
I don't think it is worth to be an exception in C.21. Such the copy-and-swap idiom should be avoided: it performs copy and swap for lvalues and move and swap for rvalues, that is not optimal. Additional calls to `clear()` and `shrink_to_fit()` after `swap()` may be required.

**Comment by Eisenwave** [comment]
> And additional calls to `clear()` and `shrink_to_fit()` after `swap()` may be required.

No, they are not. The other object gets copied or moved-from just to pass the `T` parameter. `this` then gets swapped with that parameter, whose lifetime is at most the surrounding *full-expression*.

Also, while I'm not a huge fan of this kind of assignment operator myself, if we're not going to add an exception to C.21 for this, then we should add a rule which advises against such assignment operators in the first place. The current wording just says nothing on this, making it unclear how some clang-tidy rules should be implemented.

**Comment by sergio-nsk** [comment]
If `T::T(T&&)` is missing, and `std::move(t)` is passed, then `T other` is copy constructed: possible 2x memory consumption and unexpected non empty state of `t` observed after the assignment expression.

**Comment by Eisenwave** [comment]
> If `T::T(T&&)` is missing, and `std::move(t)` is passed, then `T other` is copy constructed

Yes, but why would the move constructor be missing? That would still be a violation of C.21 and I don't seek to change that, so I don't see how this point is relevant.

**Comment by eyalroz** [comment]
> T& operator=(T) noexcept { /* ... */ }

This is not exactly a copy assignment nor quite a move assignment. But it can be used in lieu of both. So, I believe at least an exception is merited.

**Comment by purpleKarrot** [comment]
```cpp
T& T::operator=(T other) noexcept {
    swap(other);
    return *this;
}
```

This implementation is sufficient and exception safe. However, it might be confusing that invoking it may throw an exception (from the implicit invocation of the copy constructor).
